commit b834cfca29cbe1945938178581eece67d51d030b
Author: Sven Neumann <s.neumann@raumfeld.com>
Date:   Mon Nov 15 09:58:40 2010 +0100

    Fix potential crash in resource_expire()
    
    GSSDPResourceBrowser emits "resource-unavailable" before it has removed
    the resource from its cache. Now if the application changes the cache
    in response to this signal emission the code will crash.
    
    Fix this potential crash by emitting the signal after the resource
    has been removed from the cache.

diff --git a/libgssdp/gssdp-resource-browser.c b/libgssdp/gssdp-resource-browser.c
index 43816fe..fb54dbd 100644
--- a/libgssdp/gssdp-resource-browser.c
+++ b/libgssdp/gssdp-resource-browser.c
@@ -581,16 +581,23 @@ static gboolean
 resource_expire (gpointer user_data)
 {
         Resource *resource;
+        char *usn;
 
         resource = user_data;
-        
+
+        /* Steal the USN pointer from the resource as we need it for the signal
+         * emission.
+         */
+        usn = resource->usn;
+        resource->usn = NULL;
+
+        g_hash_table_remove (resource->resource_browser->priv->resources, usn);
+
         g_signal_emit (resource->resource_browser,
                        signals[RESOURCE_UNAVAILABLE],
                        0,
-                       resource->usn);
-
-        g_hash_table_remove (resource->resource_browser->priv->resources,
-                             resource->usn);
+                       usn);
+        g_free (usn);
 
         return FALSE;
 }
